jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/MassIndexingJobAvailableParameters.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.jsr352.massindexing;
+
+
+/**
+ * @author Yoann Rodiere
+ */
+public final class MassIndexingJobAvailableParameters {
```

Comment:
> MassIndexing is used in the package name, so when users use this class, they know the class for mass-indexing.

For utility classes I prefer not to rely on the package name to differentiate utilities: it messes with IDE auto-completion when you have multiple such classes with the same name, making it harder to use them. See for instance `AvailableSettings`, you have 2 or 3 of them and it's frankly annoying. On top of that, in your code you don't see the package name right away, so it may end up quite confusing.

> Available is redundant, when a parameter is exposed as a public String, it is obvious that it's available to be used. I would prefer to document the availability of different parameters in the Javadoc, and keep the class name short.

That was more to mirror the naming of the equivalent classes in Hibernate ORM / Search, but I guess it's not that important. Ok to remove this.

> In JSR 352, they use terminology @BatchProperty, so I guess JobProperties would be more consistent than JobParameters.

Not really:

 * properties are what is injected in our batch artifacts (listeners, partition mapper, etc.). 
 * parameters are what is provided by users to the job operator, and they are accessible in the job's XML definition by using `jobParameters`.
 * parameters are mapped to properties in the job's XML definition, using `<property>` tags.

I will admit I did take a shortcut when I userd `MassIndexingJobAvailableParameters` in `@BatchProperty` annotations, but it's because in our case properties and parameters have the exact same name. But what's important here is the public API: what we want to do is to expose these constants to users, so that they can refer to parameter names by themselves.

I'm not sure it's worth it, but we could create a separate utility class to clearly separate property names from parameter names. It's purely internal though, so we can always do it later.

There's another thing we could do, that would reduce the confusion between properties and parameters: we could only use `@BatchProperty` in the `JobContextSetupListener`, and make the listener responsible for putting every property in the `JobContextData`. The main advantage would be that every property would be appropriately deserialized and validated upfront, eliminating any risk of I/O error during the job execution. Maybe I should create a ticket about this?

> Singular is better than plural. Singular is shorter, and it categorizes the nature of the class. An example would be class java.time.Month, where month is singular.

We're not talking about an instantiable class here, but about a utility class that centralizes constant definitions. Our constants will not have the type `MassIndexingJobAvailableParameters`, they are strings. In this case, plural is generally preferred. See for example:

 * `org.hibernate.cfg.AvailableSettings`
 * `org.hibernate.jpa.AvailableSettings`
 * `java.nio.charset.StandardCharsets`
 * `com.google.common.base.Charsets`
 * `java.time.temporal.IsoFields`
 * `java.time.temporal.JulianFields`
 * etc.

> Here's how to final result looks like:

And here's a good example :) When you look at this code, it's quite hard to know which set of properties we're dealing with. In this case it's not that bad, since we're in the Hibernate Search code, and we know the code is about the mass indexing job, but imagine a user with `PropertyName.CACHEABLE` in his code: it's not very obvious what this property is about.

What about `MassIndexingJobParameters`? If you feel strongly about it, I guess `JobParameters` would do, but I fear we'll regret it one day, especially if we introduce another Hibernate Search job that is not related to mass indexing: we'll have to add another class named `JobParameters` in another package in the same project, which seems quite bad.